### PR TITLE
Refactored CUDA and Metal backends into extensions.

### PR DESCRIPTION
**Pull request by anicusan**
Trying to push for atomics on all KernelAbstractions.jl backends - really appreciate your help. I simplified the code structure in `ext` and `test` with the current Julia best practices I know.

The only problem with this extension-based structure is I now get:
```
┌ Warning: Circular dependency detected. Precompilation will be skipped for:
│   Base.PkgId(Base.UUID("66d79d19-2cc4-5b0b-ac7a-b340256d1ecd"), "LinearAlgebraExt")
│   Base.PkgId(Base.UUID("63c18a36-062a-441e-b654-da1e3ab1ce7c"), "KernelAbstractions")
│   Base.PkgId(Base.UUID("368c01e8-f8da-5dca-a1ea-818da1f33961"), "AtomixMetalExt")
│   Base.PkgId(Base.UUID("dde4c033-4e86-420c-a63e-0dd931031962"), "Metal")
└ @ Base.Precompilation precompilation.jl:511
```

Is this something to be fixed from this side, or from KernelAbstractions' side? The rest of the ecosystem using Atomix needs to be updated for the current `Atomix = "1"` version too, otherwise it very quickly spirals into dependency hell.

### CODE REVIEW

**Comment by anicusan** [comment]
Okay, the extensions are found and tests pass. Still, we get the circular dependency problem on the [GPU CI](https://buildkite.com/julialang/atomix-dot-jl/builds/12#01931bb1-a9c1-4b49-9f2a-3bb7904f28aa) too - I am not sure where/why/how to fix it, waiting on your input.

Afterwards, we can increment the version up to `v1.0.0` once and for all, which will then need updating across the ecosystem (y'all are really doing titanic work).

File: Project.toml [hunk 1]
```diff
@@ -1,11 +1,24 @@
...
+version = "0.1.0"
 
 [deps]
 UnsafeAtomics = "013be700-e6cd-48c3-b4a1-df204f14c38f"
 
 [compat]
+CUDA = "5"
+Metal = "1"
 UnsafeAtomics = "0.1, 0.2"
 julia = "1.6"
```
Comment (vchuravy) [inline Project.toml:13]:
```suggestion
julia = "1.10"
```

File: Project.toml [hunk 2]
```diff
@@ -1,11 +1,24 @@
 name = "Atomix"
 uuid = "a9b6321e-bd34-4604-b9c9-b65b8de01458"
 authors = ["Takafumi Arakaki <aka.tkf@gmail.com> and contributors"]
-version = "1.0"
+version = "0.1.0"
```
Comment (vchuravy) [inline Project.toml:4]: Why? 
```suggestion
version = "0.2.0"
```

**Comment by vchuravy** [comment]
Can you split it into two PRs? One for the rework to extensions and one for adding oneAPI?

File: Project.toml (hunk 2 above)
Comment (anicusan) [reply Project.toml]: Considering this messes up the compatibility with KA anyway, we might as well just bump Atomix to 1.0.0?

**Comment by christiangnrd** [comment]
The circular dependency is probably because the GPU backends depend on KernelAbstractions, which depends on Atomix. I don't know how to fix this with extensions that live in this repo.

**Comment by anicusan** [comment]
Is it a case of having to update the `[compat]` versions in tandem?

**Comment by christiangnrd** [comment]
> Is it a case of having to update the `[compat]` versions in tandem?

I don't believe so. I tested this by reverting the Atomix version to 0.1.0 to avoid the compat issues.

**Comment by christiangnrd** [comment]
I suspect this "cycle" issue may be fixed by https://github.com/JuliaLang/julia/pull/55910#issuecomment-2488843885. If true, that means that this should be working once 1.10.8 and 1.11.3 are released.

**Comment by christiangnrd** [comment]
Okay I compiled the backports-1.11 branch to test this out and I'm getting a different error.

<details>
<summary>Error:</summary>

```
│  ┌ Warning: Module Metal with build ID ffffffff-ffff-ffff-0000-ba908ccbce5a is missing from the cache.
│  │ This may mean Metal [dde4c033-4e86-420c-a63e-0dd931031962] does not support precompilation but is imported by a module that does.
│  └ @ Base loading.jl:2539
│  ┌ Error: Error during loading of extension AtomixMetalExt of Atomix, use `Base.retry_load_extensions()` to retry.
│  │   exception =
│  │    1-element ExceptionStack:
│  │    Error when precompiling module, potentially caused by a __precompile__(false) declaration in the module.
│  │    Stacktrace:
│  │      [1] _require(pkg::Base.PkgId, env::Nothing)
│  │        @ Base ./loading.jl:2601
│  │      [2] __require_prelocked(uuidkey::Base.PkgId, env::Nothing)
│  │        @ Base ./loading.jl:2386
│  │      [3] #invoke_in_world#3
│  │        @ ./essentials.jl:1089 [inlined]
│  │      [4] invoke_in_world
│  │        @ ./essentials.jl:1086 [inlined]
│  │      [5] _require_prelocked
│  │        @ ./loading.jl:2373 [inlined]
│  │      [6] _require_prelocked
│  │        @ ./loading.jl:2372 [inlined]
│  │      [7] run_extension_callbacks(extid::Base.ExtensionId)
│  │        @ Base ./loading.jl:1542
│  │      [8] run_extension_callbacks(pkgid::Base.PkgId)
│  │        @ Base ./loading.jl:1574
│  │      [9] run_package_callbacks(modkey::Base.PkgId)
│  │        @ Base ./loading.jl:1395
│  │     [10] _require_search_from_serialized(pkg::Base.PkgId, sourcepath::String, build_id::UInt128, stalecheck::Bool; reasons::Dict{String, Int64}, DEPOT_PATH::Vector{String})
│  │        @ Base ./loading.jl:2063
│  │     [11] _require(pkg::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2525
│  │     [12] __require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2386
│  │     [13] #invoke_in_world#3
│  │        @ ./essentials.jl:1089 [inlined]
│  │     [14] invoke_in_world
│  │        @ ./essentials.jl:1086 [inlined]
│  │     [15] _require_prelocked(uuidkey::Base.PkgId, env::String)
│  │        @ Base ./loading.jl:2373
│  │     [16] macro expansion
│  │        @ ./loading.jl:2312 [inlined]
│  │     [17] macro expansion
│  │        @ ./lock.jl:273 [inlined]
│  │     [18] __require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:2269
│  │     [19] #invoke_in_world#3
│  │        @ ./essentials.jl:1089 [inlined]
│  │     [20] invoke_in_world
│  │        @ ./essentials.jl:1086 [inlined]
│  │     [21] require(into::Module, mod::Symbol)
│  │        @ Base ./loading.jl:2258
│  │     [22] include
│  │        @ ./Base.jl:557 [inlined]
│  │     [23] include_package_for_output(pkg::Base.PkgId, input::String, depot_path::Vector{String}, dl_load_path::Vector{String}, load_path::Vector{String}, concrete_deps::Vector{Pair{Base.PkgId, UInt128}}, source::Nothing)
│  │        @ Base ./loading.jl:2879
│  │     [24] top-level scope
│  │        @ stdin:5
│  │     [25] eval
│  │        @ ./boot.jl:430 [inlined]
│  │     [26] include_string(mapexpr::typeof(identity), mod::Module, code::String, filename::String)
│  │        @ Base ./loading.jl:2732
│  │     [27] include_string
│  │        @ ./loading.jl:2742 [inlined]
│  │     [28] exec_options(opts::Base.JLOptions)
│  │        @ Base ./client.jl:321
│  │     [29] _start()
│  │        @ Base ./client.jl:531
│  └ @ Base loading.jl:1548
```
</details>

I'm not familiar enough with precompilation to know if that's a Julia bug or a bug in this PR.

**Pull request merged**